### DISCUSSION TITLE
CSV sniffer detects "unlikely" delimiter, column count, and header row

### DISCUSSION THREAD

**Discussion by tboddyspargo** [duckdb/duckdb#9347] [General]
### What happens?

There now seem to be a number of cases where I would argue that the sniffer is preferring a questionable guess over raising an error. In #9111, I provided a case that used to raise an exception. Recently that behavior and test case expectation were changed (https://github.com/pdet/duckdb/commit/11e85cfc818240694cb4ba74bd794ae2c81187b0). 

Here are some more examples where I feel that the quality of CSV auto detection from the Sniffer has become, in my opinion, less accurate (some of these cases seem like they should be obvious enough) and less trustworthy (raise exceptions in cases of high ambiguity).

> **NOTE:** I just want to acknowledge that the cases I'm using here are highly irregular. They have simply been our historic tool for ensuring consistent behavior when CSV files are poorly formatted. We often prefer to raise an exception rather than detect something that seems highly unlikely to be desired.

<details>
<summary>1. Detected 1 column with no header row, instead of 3 or exception</summary>

```

one,two,three
1,2,3,4
1,2,3

1,2,3
1,2,3
```

</details>

<details>
<summary>2. Detected 5 columns with no header row, instead of 3 or exception</summary>

```

one,two,three
1,2,3,4
1,2,3

1,2,3,4,5
1,2,3,4,5
```

</details>
<details>
<summary>3. Detected 1 column with no header row, instead of 3 or exception</summary>

```

one,two,three
1,2,3,4
1,2,3

1,2,3,4,5
1,2,3,4,5
1,2,3
1,2,3
1,2,3
```

</details>

Is the new expectation that a sufficiently inconsistent number of delimiters per row will result in the sniffer identifying a single column without any delimiter? Why did it decide that 5 columns was appropriate in **2** while it decided that 1 column was more likely in **3**?

In the above cases, are the data types not being used to help "weight" the sniffer in favor of `one,two,three` being the header row? I thought I remembered from documentation that was the case, but it seems to be underutilized in this case.

Would it be reasonable to expect that the CSV sniffer could identify that `one,two,three` is the header row, even if it determined that the column count was supposed to be `5` and that it, therefore, needs to append default column names to the headers where header values and delimiters are missing?

### To Reproduce

```python
import io
import duckdb


conn = duckdb.connect(':memory:')
file = io.StringIO(('\n' 'One,two,three\n' '1,2,3,4\n' '1,2,3\n' '\n' '1,2,3\n' '1,2,3\n'))
rel = conn.read_csv(file)
rel.show()

file = io.StringIO(('\n' 'One,two,three\n' '1,2,3,4\n' '1,2,3\n' '\n' '1,2,3,4,5\n' '1,2,3,4,5\n'))
rel = conn.read_csv(file)
rel.show()

file = io.StringIO(('\n' 'One,two,three\n' '1,2,3,4\n' '1,2,3\n' '\n' '1,2,3,4,5\n' '1,2,3,4,5\n' '1,2,3\n' '1,2,3\n' '1,2,3\n'))
rel = conn.read_csv(file)
rel.show()
```

### OS:

macOS_aarch64

### DuckDB Version:

v0.9.1

### DuckDB Client:

Python

### Full Name:

Tyler BoddySpargo

### Affiliation:

DataChat

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by pdet** [comment]
Hey Tyler, thanks for bringing this up. This is more of a discussion than an issue, so we should convert it to that.

The problem is that many CSV files are polluted by nature due to the flexible text format. 
We have decided on a best effort in terms of reading it, where only very unruly files would yield an exception. 

I could think of adding a `strict` mode to the sniffer, where only standard scenarios are tested, no dirty lines are skipped, and so on.  However, the CSV Reader already has so many options that I would only consider something in that direction if I see more people asking for it.

For the files you provided, playing with the delimiter, null_padding, or column definitions should produce the outcomes you desire.

**Comment by l1t1** [comment]
when bad formatted csv file detected , duckdb should raise error as early as possible
ref 
https://datatracker.ietf.org/doc/html/rfc4180
https://www.csvreader.com/csv_format.php

**Comment by tboddyspargo** [comment]
> This is more of a discussion than an issue, so we should convert it to that.

Fair enough. I don't seem to have a button for this, would you like me to just copy and paste into a discussion, or does the team have a way to convert it "with a click"?

> I could think of adding a `strict` mode to the sniffer, where only standard scenarios are tested, no dirty lines are skipped, and so on. However, the CSV Reader already has so many options that I would only consider something in that direction if I see more people asking for it.

I'm not sure I'd advocate for the added complexity of separate modes, per se. However, I'm hoping here are some reasonable ways that the accuracy can be improved a bit and/or exceptions can be raised rather than a best effort result of questionable reasonableness.

> For the files you provided, playing with the delimiter, null_padding, or column definitions should produce the outcomes you desire.

I'm afraid `delimiter` and `columns`/`names` aren't particularly reasonable workarounds for me, because I want to support "arbitrary" CSV files and I would (ideally) like to be able to rely on `duckdb` for delimiter sniffing as well as header and type detection.

It looks like `null_padding` hasn't been added to the documentation yet, but I gave it a shot with the three examples from above. Unfortunately, it did not improve the results. Instead it resulted in all examples being treated as having a single column.

I think one of the main challenges with my expectations is that if the count  per row of a candidate delimiter is not identical, then it might be technically indistinguishable from a file with null delimiters (meaning: a single column).

Some CSV sniffers use a more [probabilistic approach](https://github.com/python/cpython/blob/0102faf0cac107fd4d0fde9902560c7cf0a3be9d/Lib/csv.py#L289-L303), which I'm admittedly biased toward due to habit/familiarity. Would you consider an approach like that for delimiter detection? The tradeoff (aside from the additional logic/complexity) would be a wider margin of error for single-column datasets to accidentally get attributed with a delimiter. However, I would argue that mitigating with support for quoted values might provide a really nice balance and meaningful improvement over the current state.

To summarize, I am proposing/requesting:
- Prefer identifying a delimiter if it reaches some threshold of consistency across sampled rows [like this](https://github.com/python/cpython/blob/0102faf0cac107fd4d0fde9902560c7cf0a3be9d/Lib/csv.py#L343-L349), null/newline delimiter means single-column, raise exception if neither criteria is met.
- Support quoted values (ignoring delimiters that occur within quotes) so that there's a reliable pathway for single-column files to be successfully "detected" as such.

Thanks for your consideration!

**Reply by pdet** [reply]
I'm all in to increase reliability, and we already have some rules for that.
The current [dialect detection](https://github.com/duckdb/duckdb/blob/main/src/execution/operator/csv_scanner/sniffer/dialect_detection.cpp#L140) gives priority to dialects that yield more columns with consistent rows (i.e., tuples that produce the same number of columns under that possible dialect candidate) as our prime candidate. This already takes into consideration delimiter values within quotes.

There are two reasons, from what you explained, that I think throwing errors is not trivial:

Note Skipping: After detecting it, in a later phase, we also detect if there is a column mismatch on the initial rows. If so, we consider these as notes (some CSV files start off with that), and we later skip them.
Empty Line Handling: Empty lines have a different meaning whether it's a single-column CSV file or not. If it's a single column, they represent a NULL value; if it's a multi-column file, we skip them. We have seen many CSV files with this pattern as well.
My plan to move forward is to add the [Pollock paper](https://dl.acm.org/doi/10.14778/3594512.3594518) as a benchmark and fuzzing tool, with the goal of reading as many CSV files as we can. At that point, we will also have a more concrete metric of accuracy, since we have a proper benchmark to measure what is "correct". So, I'm not sure it will go towards the outcome you desire.

I must say none of this is currently in my plans for the next release.

**Reply by tboddyspargo** [reply]
I really appreciate you exploring this with me, @pdet. I think I've been playing catch up to the deep thinking you've clearly done about `duckdb`'s sniffing capabilities. I'm coming around to the conclusion that the recent changes are very welcome and a meaningful improvement.

> It looks like `null_padding` hasn't been added to the documentation yet, but I gave it a shot with the three examples from above. Unfortunately, it did not improve the results. Instead it resulted in all examples being treated as having a single column.

I need to make a correction. I don't think I was testing with the version of `duckdb` that I thought I was... `null_padding` did improve my results in some (perhaps most) situations. The only problematic scenarios when `null_padding=true` seem to be: 
- CSVs with multiple columns where initial rows contain notes that are intended/expected to be considered "pre-header" rows
- CSVs with multiple columns where initial rows contain delimiters, but only NULL values that are intended/expected to be considered "pre-header" rows

<details><summary>Example 1</summary>
<p>

```

This is a note.
a,b,c
1,2,3
1,2,3
1,2,3
```

</p>
</details> 

<details><summary>Example 2</summary>
<p>

```
,,
a,b,c
1,2,3
1,2,3
1,2,3
```

</p>
</details> 

The two scenarios share a lot of commonalities, but seem like tricky places to navigate for "header detection" when `null_padding=true` is requested.

> Note Skipping: After detecting it, in a later phase, we also detect if there is a column mismatch on the initial rows. If so, we consider these as notes (some CSV files start off with that), and we later skip them.

That makes sense. Note skipping has been a challenge for us in the past, too. From some additional testing, it looks like `duckdb`'s Note skipping logic works well for my small sample size of cases when `header=True,null_padding=false`, however an initial "note row" will always be treated as the header row when `header=True,null_padding=true` and it contains delimiters. I don't consider this a big issue, but it does mean that we will struggle to support CSVs with "initial note rows" (or pre-header rows that contain any delimiters at all) if we use `null_padding=True`. I'm not sure if you see any way to provide better consistency there...

> Empty Line Handling: Empty lines have a different meaning whether it's a single-column CSV file or not. If it's a single column, they represent a NULL value; if it's a multi-column file, we skip them. We have seen many CSV files with this pattern as well.
> My plan to move forward is to add the [Pollock paper](https://dl.acm.org/doi/10.14778/3594512.3594518) as a benchmark and fuzzing tool, with the goal of reading as many CSV files as we can.

I was realizing this over the weekend as I expanded my testing. I think your plan is a great one and I agree with the vision.

**Reply by pdet** [reply]
> I really appreciate you exploring this with me, @pdet. I think I've been playing catch up to the deep thinking you've clearly done about `duckdb`'s sniffing capabilities. I'm coming around to the conclusion that the recent changes are very welcome and a meaningful improvement.

Thanks for the message. I'm glad you agree with the changes :-)

> > It looks like `null_padding` hasn't been added to the documentation yet, but I gave it a shot with the three examples from above. Unfortunately, it did not improve the results. Instead it resulted in all examples being treated as having a single column.
> 
> I need to make a correction. I don't think I was testing with the version of `duckdb` that I thought I was... `null_padding` did improve my results in some (perhaps most) situations. The only problematic scenarios when `null_padding=true` seem to be:

Thanks for pointing out this option was missing from the documentation, I've PR-ed it yesterday, and it should be there now.
> * CSVs with multiple columns where initial rows contain notes that are intended/expected to be considered "pre-header" rows
> * CSVs with multiple columns where initial rows contain delimiters, but only NULL values that are intended/expected to be considered "pre-header" rows
> 
> Example 1
> ```
> 
> This is a note.
> a,b,c
> 1,2,3
> 1,2,3
> 1,2,3
> ```
> 
> Example 2
> ```
> ,,
> a,b,c
> 1,2,3
> 1,2,3
> 1,2,3
> ```
> 
> The two scenarios share a lot of commonalities, but seem like tricky places to navigate for "header detection" when `null_padding=true` is requested.
> 
> > Note Skipping: After detecting it, in a later phase, we also detect if there is a column mismatch on the initial rows. If so, we consider these as notes (some CSV files start off with that), and we later skip them.
> 
> That makes sense. Note skipping has been a challenge for us in the past, too. From some additional testing, it looks like `duckdb`'s Note skipping logic works well for my small sample size of cases when `header=True,null_padding=false`, however an initial "note row" will always be treated as the header row when `header=True,null_padding=true` and it contains delimiters. I don't consider this a big issue, but it does mean that we will struggle to support CSVs with "initial note rows" (or pre-header rows that contain any delimiters at all) if we use `null_padding=True`. I'm not sure if you see any way to provide better consistency there...

I agree that this is problematic. I plan to improve header detection drastically. There is a list of approximately 1M common header names. The basic idea is to include a bloom filter with these header names and, if matches happen on the rows before the data, to use that row as a header.
